engine/interpreter/src/main/scala/pl/touk/nussknacker/engine/marshall/ProcessMarshaller.scala

```diff
@@ -121,9 +123,41 @@ class ProcessMarshaller(implicit
 }
 
 object ProcessMarshaller {
-  def emptyOptionCodec[T]: CodecJson[Option[T]] = CodecJson.apply[Option[T]](_ => jEmptyObject, c => DecodeResult.ok(Option.empty[T]))
 
-  val additionalNodeDataFieldsCodec: CodecJson[Option[node.UserDefinedAdditionalNodeFields]] = emptyOptionCodec
+  val additionalNodeDataFieldsCodec: CodecJson[Option[UserDefinedAdditionalNodeFields]] =
+    derivedTypeOrNoneCodec[UserDefinedAdditionalNodeFields, NodeAdditionalFields]
+
+  val additionalProcessFieldsCodec: CodecJson[Option[UserDefinedProcessAdditionalFields]] = {
+    val derivedCodec = derivedTypeOrNoneCodec[UserDefinedProcessAdditionalFields, ProcessAdditionalFields]
+    CodecJson.derived(
+      derivedCodec.Encoder,
+      DecodeJson(c => {
+        val withEmptyWhenMissing = (c --\ "metaData" --\ "additionalFields")
+          .reattempt
+          .withFocus(_.withObject(additionalFields => {
+            val withGroups = additionalFields("groups")
```

Comment:
It is a little bit confusing to read at the moment. Consider something similar to `additionalFields + ("groups" -> additionalFields("groups").getOrElse(jEmptyArray))` to make it more readable?

Also, I don't like the idea of accessing fields by their names but I don't know if you can easily get around it.

Reply:
I decided to decode each field explicitly. In case of empty option, field is decode to empty set/empty map.

Reply:
Alright, this looks good to me. Unfortunatelly I don't have the power to approve it. @mproch @nadberezny?